File: rust-sdk/golem-rust/src/uuid.rs
// Copyright 2024 Golem Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::bindings::golem::api::host::ComponentId;
use std::fmt;
use std::fmt::{Display, Formatter};
use std::str::FromStr;
use uuid::Uuid;

impl From<crate::bindings::golem::api::host::Uuid> for Uuid {
    fn from(uuid: crate::bindings::golem::api::host::Uuid) -> Self {
        Uuid::from_u64_pair(uuid.high_bits, uuid.low_bits)
    }
}

impl From<Uuid> for crate::bindings::golem::api::host::Uuid {
    fn from(value: Uuid) -> Self {
        let (high_bits, low_bits) = value.as_u64_pair();
        Self {
            high_bits,
            low_bits,
        }
    }
}

impl From<Uuid> for crate::bindings::golem::api::host::ComponentId {
    fn from(value: Uuid) -> Self {
        Self { uuid: value.into() }
    }
}

impl From<crate::bindings::golem::api::host::ComponentId> for Uuid {
    fn from(value: crate::bindings::golem::api::host::ComponentId) -> Self {
        value.uuid.into()
    }
}

impl FromStr for crate::bindings::golem::api::host::ComponentId {
    type Err = uuid::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Uuid::parse_str(s)?.into())
    }
}

impl Display for crate::bindings::golem::api::host::ComponentId {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", &self.uuid)
    }
}

impl Display for crate::bindings::golem::api::host::WorkerId {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}/{}", self.component_id, self.worker_name)
    }
}

impl FromStr for crate::bindings::golem::api::host::WorkerId {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<&str> = s.split('/').collect();
        if parts.len() == 2 {
            let component_id = ComponentId::from_str(parts[0])
                .map_err(|_| format!("invalid component id: {s} - expected uuid"))?;
            let worker_name = parts[1].to_string();
            Ok(Self {
                component_id,
                worker_name,
            })
        } else {
            Err(format!(
                "invalid worker id: {s} - expected format: <component_id>/<worker_name>"
            ))
        }
    }
}

impl Display for crate::bindings::golem::api::host::PromiseId {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}/{}", self.worker_id, self.oplog_idx)
    }
}

impl FromStr for crate::bindings::golem::api::host::PromiseId {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<&str> = s.split('/').collect();
        if parts.len() == 2 {
            let worker_id = crate::bindings::golem::api::host::WorkerId::from_str(parts[0])
                .map_err(|_| {
                    format!(
                        "invalid worker id: {s} - expected format: <component_id>/<worker_name>"
                    )
                })?;
            let oplog_idx = parts[1]
                .parse()
                .map_err(|_| format!("invalid oplog index: {s} - expected integer"))?;
            Ok(Self {
                worker_id,
                oplog_idx,
            })
        } else {
            Err(format!(
                "invalid promise id: {s} - expected format: <worker_id>/<oplog_idx>"
            ))
        }
    }
}

File: sdks/scala/model/src/main/scala/golem/Uuid.scala
/*
 * Copyright 2024-2026 Golem Cloud
 *
 * Licensed under the Golem Source License v1.1 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://license.golem.cloud/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golem

import zio.blocks.schema.Schema

/**
 * Cross-platform UUID representation used by the Scala SDK.
 */
final case class Uuid(highBits: BigInt, lowBits: BigInt)

object Uuid {
  implicit val schema: Schema[Uuid] = Schema.derived

  def random(): Uuid = {
    val high = (randomUnsigned64() & ~BigInt(0xf000)) | BigInt(0x4000)
    val low  = (randomUnsigned64() & ~(BigInt(0xc000) << 48)) | (BigInt(0x8000) << 48)
    Uuid(high, low)
  }

  private def randomUnsigned64(): BigInt = {
    val high = (scala.math.random() * 4294967296.0).toLong
    val low  = (scala.math.random() * 4294967296.0).toLong
    (BigInt(high) << 32) | BigInt(low)
  }

  def toStandardString(uuid: Uuid): String = {
    val hi = uuid.highBits
    val lo = uuid.lowBits
    val p1 = ((hi >> 32) & 0xffffffffL).toLong
    val p2 = ((hi >> 16) & 0xffffL).toLong
    val p3 = (hi & 0xffffL).toLong
    val p4 = ((lo >> 48) & 0xffffL).toLong
    val p5 = lo & 0xffffffffffffL
    f"$p1%08x-$p2%04x-$p3%04x-$p4%04x-$p5%012x"
  }

  def fromStandardString(s: String): Either[String, Uuid] = {
    val parts = s.split('-')
    if (parts.length != 5) Left(s"Invalid UUID string: $s")
    else
      try {
        val p1 = BigInt(parts(0), 16)
        val p2 = BigInt(parts(1), 16)
        val p3 = BigInt(parts(2), 16)
        val p4 = BigInt(parts(3), 16)
        val p5 = BigInt(parts(4), 16)
        val hi = (p1 << 32) | (p2 << 16) | p3
        val lo = (p4 << 48) | p5
        Right(Uuid(hi, lo))
      } catch {
        case _: NumberFormatException => Left(s"Invalid UUID string: $s")
      }
  }
}
